### PR TITLE
Receive component refactor to get new addresses

### PR DESCRIPTION
**Pull request by banklesss**
https://emurgo.atlassian.net/browse/YOMO-399

### CODE REVIEW

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 1]
```diff
@@ -70,7 +70,7 @@ import {toCachedTx, TransactionCache} from './shelley/transactionCache'
 import {yoroiSignedTx} from './signedTx'
 import {NetworkId, WalletImplementationId, WalletInterface, YoroiProvider} from './types'
 import {yoroiUnsignedTx} from './unsignedTx'
-import {generateUtxoStorage} from './utxoStorage'
+import {makeUtxoStorage} from './utxoStorage'
```
Comment (banklesss) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:83]: just normalising the names https://github.com/Emurgo/yoroi-mobile/pull/2209/files#diff-35be0faa23c723910b92f1d9db6b12ae6568381be0ba00ac07d2c5d3a08c921cR94

File: src/yoroi-wallets/Wallet.ts [hunk 2]
```diff
@@ -307,37 +306,6 @@ export class Wallet {
     this.notify({type: 'state', state: this.state})
   }
 
-  canGenerateNewReceiveAddress() {
```
Comment (wolverineks) [inline src/yoroi-wallets/Wallet.ts]: why did you move this?

Comment (banklesss) [reply src/yoroi-wallets/Wallet.ts]: The methods has been moved from `walletManager` too. There, one method invoked `wallet.save()`, so considering `Wallet` class is going to be removed I tried to avoid `super` again.

Comment (stackchain) [reply src/yoroi-wallets/Wallet.ts]: I think it's correct for the impl to be inside the specialized class, but I would expect to have these methods in the wallet interface.

**Comment by wolverineks** [comment]
which changes are cleanups and which changes change behavior?

File: src/legacy/actions.ts [hunk 3]
```diff
@@ -127,14 +127,6 @@ export const setupHooks = () => (dispatch: Dispatch<any>) => {
   walletManager.subscribeServerSync((status) => dispatch(_setServerStatus(status)))
 }
 
-export const generateNewReceiveAddress = () => async (_dispatch: Dispatch<any>) => {
```
Comment (banklesss) [inline src/legacy/actions.ts:130]: cleanup

File: src/yoroi-wallets/walletManager.ts [hunk 4]
```diff
@@ -193,25 +193,6 @@ export class WalletManager {
     }
   }
 
-  // ========== UI state ============= //
```
Comment (banklesss) [inline src/yoroi-wallets/walletManager.ts:195]: cleanup. Moved to `ShelleyWallet`

File: src/yoroi-wallets/cardano/types.ts [hunk 5]
```diff
@@ -133,14 +135,6 @@ export interface WalletInterface {
 
   tryDoFullSync(): Promise<void>
 
-  // =================== state/UI =================== //
```
Comment (banklesss) [inline src/yoroi-wallets/cardano/types.ts:145]: cleanup. Removing UI references

File: src/Receive/ReceiveScreen.tsx [hunk 6]
```diff
@@ -2,33 +2,26 @@ import _ from 'lodash'
...
   const wallet = useSelectedWallet()
-  const receiveAddresses = wallet.externalAddresses.slice(0, wallet.numReceiveAddresses)
+  const receiveAddresses = useReceiveAddresses(wallet)
   const addressLimitReached = wallet.canGenerateNewReceiveAddress() == false
 
   const currentAddress = _.last(receiveAddresses)
 
-  const dispatch = useDispatch()
   React.useEffect(() => {
-    dispatch(generateNewReceiveAddressIfNeeded())
```
Comment (banklesss) [inline src/Receive/ReceiveScreen.tsx:24]: cleanup

File: src/Receive/ReceiveScreen.tsx [hunk 7]
```diff
@@ -2,33 +2,26 @@ import _ from 'lodash'
...
-  }, [dispatch])
-
-  // This is here just so that we can properly monitor changes and fire
-  // generateNewReceiveAddressIfNeeded()
-  React.useEffect(() => {
-    dispatch(generateNewReceiveAddressIfNeeded())
```
Comment (banklesss) [inline src/Receive/ReceiveScreen.tsx:30]: cleanup

File: src/Receive/ReceiveScreen.tsx [hunk 8]
```diff
@@ -2,33 +2,26 @@ import _ from 'lodash'
...
 import {useSelectedWallet} from '../SelectedWallet'
 import {COLORS} from '../theme'
+import {YoroiWallet} from '../yoroi-wallets'
 import {AddressDetail} from './AddressDetail'
 import {UnusedAddresses, UsedAddresses} from './Addresses'
 
 export const ReceiveScreen = () => {
   const strings = useStrings()
   const wallet = useSelectedWallet()
-  const receiveAddresses = wallet.externalAddresses.slice(0, wallet.numReceiveAddresses)
```
Comment (banklesss) [inline src/Receive/ReceiveScreen.tsx:17]: Bug. Component is unaware of addresses changes. Fix: https://github.com/Emurgo/yoroi-mobile/pull/2209/files#diff-c94a2ef1655f7c1de12a45bf67350d7a4ab1c6d8381ae79cc03441ed7d1854b6R64

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 9]
```diff
@@ -512,6 +516,42 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
...
+    if (!this.canGenerateNewReceiveAddress()) return false
+
+    this.updateState({
+      lastGeneratedAddressIndex: this.state.lastGeneratedAddressIndex + 1,
+    })
+
+    // note: don't await on purpose
+    this.save()
+
+    this.notify({type: 'addresses', addresses: this.receiveAddresses})
```
Comment (banklesss) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:601]: Added this so receive component is notified when a new address is generated.

File: src/Receive/Addresses.tsx [hunk 10]
```diff
@@ -181,15 +181,15 @@ const useAddressIndex = (address: string) => {
 
 const useUnusedAddresses = () => {
```
Comment (banklesss) [inline src/Receive/Addresses.tsx]: Should this be a `ShelleyWallet` getter?

File: src/Receive/Addresses.tsx [hunk 11]
```diff
@@ -181,15 +181,15 @@ const useAddressIndex = (address: string) => {
...
   const wallet = useSelectedWallet()
-  const receiveAddresses = wallet.externalAddresses.slice(0, wallet.numReceiveAddresses)
+  const receiveAddresses = wallet.receiveAddresses
   const isUsedAddressIndex = wallet.isUsedAddressIndex
 
   return receiveAddresses.filter((address) => isUsedAddressIndex[address] !== true)
 }
 
 const useUsedAddresses = () => {
```
Comment (banklesss) [inline src/Receive/Addresses.tsx]: Should this be a ShelleyWallet getter?

File: src/Receive/ReceiveScreen.tsx [hunk 12]
```diff
@@ -68,6 +61,12 @@ export const ReceiveScreen = () => {
   )
 }
 
+const useReceiveAddresses = (wallet: YoroiWallet) => {
```
Comment (wolverineks) [inline src/Receive/ReceiveScreen.tsx]: move to hooks

File: src/Receive/Addresses.tsx [hunk 13]
```diff
@@ -181,15 +181,15 @@ const useAddressIndex = (address: string) => {
 
 const useUnusedAddresses = () => {
   const wallet = useSelectedWallet()
-  const receiveAddresses = wallet.externalAddresses.slice(0, wallet.numReceiveAddresses)
+  const receiveAddresses = wallet.receiveAddresses
```
Comment (wolverineks) [inline src/Receive/Addresses.tsx]: const receiveAddresses = useWallet(wallet, "addresses")
useWallet(wallet, "transactions")

File: src/Receive/Addresses.tsx [hunk 14]
```diff
@@ -181,15 +181,15 @@ const useAddressIndex = (address: string) => {
...
   const isUsedAddressIndex = wallet.isUsedAddressIndex
 
   return receiveAddresses.filter((address) => isUsedAddressIndex[address] !== true)
 }
 
 const useUsedAddresses = () => {
   const wallet = useSelectedWallet()
-  const receiveAddresses = wallet.externalAddresses.slice(0, wallet.numReceiveAddresses)
+  const receiveAddresses = wallet.receiveAddresses
```
Comment (wolverineks) [inline src/Receive/Addresses.tsx]: const receiveAddresses = useWallet(wallet, "addresses")
useWallet(wallet, "transactions")

**Review by stackchain — APPROVED** [review]

File: src/Receive/Addresses.tsx [hunk 15]
```diff
@@ -8,11 +8,12 @@ import {useReceiveAddresses, useWallet} from '../hooks'
 import {isEmptyString} from '../legacy/utils'
 import {useSelectedWallet} from '../SelectedWallet'
 import {COLORS} from '../theme'
+import {Address as AddressType} from '../yoroi-wallets/types'
```
Comment (wolverineks) [inline src/Receive/Addresses.tsx:11]: why rename the type?

**Review by wolverineks — APPROVED** [review]
